cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tethering/TetheringServerHandler.java

```diff
@@ -0,0 +1,265 @@
+      LOG.warn("Topic {} already exists", topicId);
+    } catch (IOException e) {
+      LOG.error("Failed to create topic {}", topicId, e);
+      throw e;
+    }
+
+    try {
+      store.getPeer(tetherRequest.getPeer());
+      // Peer is already configured, treat this as a no-op.
+      responder.sendStatus(HttpResponseStatus.OK);
+      return;
+    } catch (PeerNotFoundException ignored) {
+    }
+
+    // We don't need to keep track of the client metadata on the server side.
+    PeerMetadata peerMetadata = new PeerMetadata(tetherRequest.getNamespaceAllocations(), Collections.emptyMap());
+    // We don't store the peer endpoint on the server side because the connection is initiated by the client.
+    PeerInfo peerInfo = new PeerInfo(tetherRequest.getPeer(), null, TetheringStatus.PENDING, peerMetadata);
+    try {
+      store.addPeer(peerInfo);
+    } catch (Exception e) {
+      try {
+        messagingService.deleteTopic(topicId);
+      } catch (Exception ex) {
+        e.addSuppressed(ex);
+      }
+      throw new IOException("Failed to create tethering with peer " + tetherRequest.getPeer(), e);
+    }
+    responder.sendStatus(HttpResponseStatus.OK);
+  }
+
+  /**
+   * Accepts/rejects the tethering request.
+   */
+  @POST
+  @Path("/tethering/connections/{peer}")
+  public void tetherAction(HttpRequest request, HttpResponder responder, @PathParam("peer") String peer,
+                           @QueryParam("action") String action)
+    throws NotImplementedException, BadRequestException, PeerNotFoundException, IOException {
+    TetheringStatus tetheringStatus;
+    switch (action) {
+      case "accept":
+        tetheringStatus = TetheringStatus.ACCEPTED;
+        break;
+      case "reject":
+        tetheringStatus = TetheringStatus.REJECTED;
+        break;
+      default:
+        throw new BadRequestException(String.format("Invalid action: %s", action));
+    }
+    updateTetherStatus(responder, peer, tetheringStatus);
+  }
+
+  private void checkTetheringServerEnabled() throws NotImplementedException {
+    if (!cConf.getBoolean(Constants.Tethering.TETHERING_SERVER_ENABLED)) {
+      throw new NotImplementedException("Tethering is not enabled");
+    }
+  }
+
+  private void updateTetherStatus(HttpResponder responder, String peer, TetheringStatus newStatus)
+    throws NotImplementedException, PeerNotFoundException, IOException {
+    checkTetheringServerEnabled();
+    PeerInfo peerInfo = store.getPeer(peer);
+    if (peerInfo.getTetheringStatus() == TetheringStatus.PENDING) {
+      store.updatePeerStatus(peerInfo.getName(), newStatus);
+    } else {
+      LOG.info("Cannot update tether state to {} as current state state is {}",
+               newStatus, peerInfo.getTetheringStatus());
+    }
+    responder.sendStatus(HttpResponseStatus.OK);
+  }
+
+  // Currently unused.
+  // TODO: Send RUN_PIPELINE message to peer.
+  private void publishMessage(TopicId topicId, @Nullable TetheringControlMessage message)
```

Comment:
What is the future usage of this method?

Reply:
it was intended for publishing RUN_PIPELINE messages that will be sent in control response. I have removed it - will add this code when we are implementing the tethering provisioner.